### PR TITLE
tRPC 10 Support

### LINKED ISSUE #52: Invalid router code generation
### Bug description

TRPC newer versions have deprecated older way to creating routers and procedures. 

Migration guide: https://trpc.io/docs/migrate-from-v9-to-v10

### How to reproduce

Just add plugin and run generator

### Expected behavior

Should generate valid code for TRPC 10

### Prisma information

Any

### Environment & setup

- OS: Linux
- Database: PostgreSQL
- Node.js version: 18

### Prisma Version

```
prisma                  : 4.7.1
@prisma/client          : 4.7.1
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Migration Engine        : migration-engine-cli 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/migration-engine-debian-openssl-3.0.x)
Introspection Engine    : introspection-core 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/introspection-engine-debian-openssl-3.0.x)
Format Binary           : prisma-fmt 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/prisma-fmt-debian-openssl-3.0.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.7.1-1.272861e07ab64f234d3ffc4094e32bd61775599c
Default Engines Hash    : 272861e07ab64f234d3ffc4094e32bd61775599c
Studio                  : 0.477.0

```

### PR DESCRIPTION
**Pull request by mwarger**
fixes #52

### Description

Adds support for tRPC v10

### CODE REVIEW

**Comment by omar-dulaimi** [comment]
Hey @mwarger thanks for the PR!
I thought tRPC v10 is stable right now. So, if that was the case, is there any reason to keep this PR behind a flag?

**Comment by mwarger** [comment]
> Hey @mwarger thanks for the PR!
> I thought tRPC v10 is stable right now. So, if that was the case, is there any reason to keep this PR behind a flag?

Just considering people that may be using this but not upgraded yet. 

I also did this because the middleware and shield features aren't implemented. Wasn't sure how to do that yet.


Happy to remove it, or we can later.

**Comment by JC-Perso** [comment]
Thx for your work ! 🚀

**Comment by innellea** [comment]
Any news about this PR @mwarger , @omar-dulaimi  ?

**Comment by sparkinson-vhs** [comment]
> Any news about this PR @mwarger , @omar-dulaimi ?

I second that. We're currently just starting new project with trpc v10 and would like to use this generator as well.

**Comment by mwarger** [comment]
@omar-dulaimi @JC-Perso @innellea @sparkinson-vhs This could be published behind a different tag if this is useful - would be nice to have other people test and provide feedback. I'm not familiar with trpc shield so I did not implement that.  I also skipped the middleware because I didn't need it.  I was just trying to get the basics out to start with which is why I put it behind a config option.  I can try to make this more feature complete this week.

**Comment by omar-dulaimi** [comment]
Hey everybody, I have been a little bit busy lately. But I promise I'll upgrade the shield and shield generator to v10, so we could merge this PR.

@mwarger I think this project will support v10 only going forward, so there's no need for the flag.

**Comment by mwarger** [comment]
I will take out the flag.

**Comment by sparkinson-vhs** [comment]
Great! Many thanks to you both.

**Comment by omar-dulaimi** [comment]
Hey all, `trpc-shield` has been updated to support `tRPC v10` now starting with `trpc-shield v0.2.0`(latest version is `0.2.1`).
The other project `prisma-trpc-shield-generato` requires no changes at all.

Which means @mwarger can proceed with uncommenting the sections related to them after updating the `trpc-shield` dependency to latest version.

I will add a few comments to this PR since @mwarger is going to need them to use the shield.

We can release as soon as work on this PR is done :)

File: prisma/schema.prisma [hunk 1]
```diff
@@ -6,6 +6,10 @@ generator trpc {
   provider          = "node ./lib/generator.js"
   isGenerateSelect  = true
   isGenerateInclude = true
+  withMiddleware = false
+  withShield     = false
+  contextPath    = "../../../../example/context"
+  v10           = true
```
Comment (omar-dulaimi) [inline prisma/schema.prisma:12]: No need for the flag anymore. So all changes related below should happen in the original functions(without the v10 duplication)

File: src/helpers.ts [hunk 2]
```diff
@@ -73,6 +83,39 @@ export function generateBaseRouter(sourceFile: SourceFile, config: Config) {
...
+  export const publicProcedure = t.procedure;`);
+
+  const middlewares = [];
+  if (config.withMiddleware) {
+    middlewares.push(/* ts */ `
+    .use(({ ctx, next }) => {
+      console.log("inside middleware!")
+      return next();
+    })`);
+  }
```
Comment (omar-dulaimi) [inline src/helpers.ts:105]: Here's how the shield will be used:

```ts
export const permissionsMiddleware = t.middleware(permissions)

export const shieldedProcedure = t.procedure.use(permissionsMiddleware)

export const publicProcedure = t.procedure

```

Then inside routers:

```ts
import { shieldedProcedure, t } from './helpers/createRouter'

export const usersRouter = t.router({
  createOneUser: shieldedProcedure.input(UserCreateOneSchema).mutation(async ({ ctx, input }) => {
    const createOneUser = await ctx.prisma.user.create(input)
    return createOneUser
  }),
});
```

If with middleware:

```ts
export const shieldedProcedure = t.procedure.use(globalMiddleware).use(permissionsMiddleware)

```

File: src/helpers.ts [hunk 3]
```diff
@@ -73,6 +83,39 @@ export function generateBaseRouter(sourceFile: SourceFile, config: Config) {
...
+
+  sourceFile.addStatements(/* ts */ `
+  export const t = trpc.initTRPC.context<Context>().create();
+  `);
+
+  sourceFile.addStatements(/* ts */ `
+  export const publicProcedure = t.procedure;`);
+
+  const middlewares = [];
+  if (config.withMiddleware) {
```
Comment (omar-dulaimi) [inline src/helpers.ts:99]: As for the middleware:

```ts
export const globalMiddleware = t.middleware(async ({ ctx, next }) => {
  console.log('inside middleware!')
  return next()
})
```

File: README.md [hunk 4]
```diff
@@ -22,17 +22,20 @@
...
   <a href="https://www.buymeacoffee.com/omardulaimi">
-    <img src="https://cdn.buymeacoffee.com/buttons/default-black.png" alt="Buy Me A Coffee" height="41" width="174">
+    <img src="https://cdn.buymeacoffee.com/buttons/default-black.png" alt="Buy Me A Coffee" height="41" width="174"/>
   </a>
 </p>
 
 ## Table of Contents
 
 - [About The Project](#about-the-project)
 - [Supported Prisma Versions](#supported-prisma-versions)
```
Comment (omar-dulaimi) [inline README.md:32]: We also need another section for supported tRPC versions

**Review by omar-dulaimi — CHANGES_REQUESTED** [review]
Please let me know if you need anything else :)

**Comment by omar-dulaimi** [comment]
Hey @mwarger 
I see that you have removed the flag. Could you also resolve my comments above so we could merge this?

**Comment by mwarger** [comment]
@omar-dulaimi Changes pushed.  I have not used and am not familiar with trpc shield, so I gave it my best shot.  I have not tested it.  Please review and feel free to push any changes if needed.  Middleware should be good too, I think, but I haven't had a chance to test it.  I just wanted to get the basics out the door based on your comments.  

I also updated dependency versions and updated the README.

**Comment by omar-dulaimi** [comment]
Looks good to me, thank you @mwarger 
Releasing soon

**Comment by omar-dulaimi** [comment]
Hey all,
Released in https://github.com/omar-dulaimi/prisma-trpc-generator/releases/tag/0.8.0
Thanks for your work @mwarger.

Guys as you know this is an open source project that I work on in my free time(when I get any). So I really appreciate any help that comes to improve and maintain it.

Also, don't forget to star it and maybe even sponsor it if you find it useful at work :) 

Thank you

**Pull request merged**